platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,709 @@
+have to consider reconnect at all.
+
+## Modules
+
+### Gossip
+
+The Gossip module is responsible for gossiping messages between neighbors. The actual gossip implementation is not
+described here, except to say that it will be possible to define and implement both event-aware and event-agnostic
+gossip implementations either to a fully connected network or where the set of neighbors is a subset of the whole. This
+document does not dictate whether raw TCP, UDP, HTTP2, gRPC, or other network protocols are used. This will be left to
+the design documents for Gossip.
+
+![Gossip](gossip-module.png)
+
+Gossip is the only part of Consensus that communicates over the network with gossip neighbors. When Gossip is
+initialized, it is supplied a roster. This roster contains the full set of nodes participating in gossip, along with
+their metadata such as RSA signing keys, IP addresses, and so forth. The Gossip module decides which neighbors to gossip
+with (using whatever algorithm it chooses).
+
+#### Events
+
+Gossip is event-oriented, meaning that it is given events to gossip, and emits events it receives through gossip. An
+implementation of Gossip could be based on a lower-level implementation based on bytes, but at the module level, it
+works in terms of events.
+
+When the Gossip module receives events through gossip, it *may* choose to perform some deduplication before sending
+them to Event Intake, but it is not required to do so.
+
+Some gossip algorithms send events in *topological order*. A neighbor may still receive events out of order, because
+different events may arrive from different neighbors at different times. Events received by Gossip are not immediately
+retransmitted to its neighbors. **An honest node will only send valid events through gossip**. If invalid events are
+ever received, you may know the node that sent them to you is dishonest. Validating events increases latency at each
+"hop", but allows us to identify dishonest gossip neighbors and discipline them accordingly. For this reason, events
+received by Gossip are sent to Event Intake, which eventually send valid, ordered events _back_ to Gossip for
+redistribution.
+
+During execution, for all nodes that are online and able to keep up, events are received "live" and processed
+immediately and re-gossiped. However, if a node is offline and then comes back online, or is starting back up after
+reconnect, it may be missing events. In this case, the node will need to ask its neighbors for any events it is missing.
+
+For this reason, every honest node needs to buffer some events, so when its neighbors ask it for events, it is able to
+send them. The Gossip module **may** cache all non-expired events, but **must** cache all non-ancient events.
+Non-expired events are crucial to support "catch up" after reconnect, where a node, after reconnecting, may still be
+several minutes behind its neighbors, and catches up by receiving those older events through gossip.
```

Comment:
Even though the system always goes through a catch up phase after a reconnect, this catching up is algorithmically indistinguishable from normal gossip when one peer falls a little behind. This sort of catch up happens all the time without reconnects happening. I think it would be a mistake to label this sort of catch up as a part of the reconnect algorithm, which up until now very specifically refers to the state transfer mechanism.

Reply:
You're correct in your description of how the system works today. On the other hand, this design proposal explores ideas of how we want the system to work in the future. And I'm just trying to look at this reconnect/catching-up process from a different perspective.

1. Theoretically, if nodes have sufficient memory/storage, then the whole reconnect could be reduced to the algorithm that we use for this catching-up phase.
2. Similarly, if nodes are sufficiently fast and have fast links between each another, then the catching-up could be replaced with the reconnect.
3. Finally, as mentioned earlier, at the high level the purpose of both the processes is exactly the same - to catch up to the latest state.

So there seems to be a room for separating responsibilities and simplifying components of the system. Specifically, a component responsible for the "catching-up" phase could be a separate entity in the architecture diagram and/or the code-base.

Reply:
> So there seems to be a room for separating responsibilities and simplifying components of the system. Specifically, a component responsible for the "catching-up" phase could be a separate entity in the architecture diagram and/or the code-base.

Separating catch up logic from the rest of the gossip logic would result in a drastic increase in the complexity of the gossip subsystem. There isn't a binary change in the mode of operation between a node gossiping normally and a node catching up. Even though we conceptually think of a node as being in a "catch up phase", in reality it's just the normal gossip algorithm in operation.

> So there seems to be a room for separating responsibilities and simplifying components of the system. Specifically, a component responsible for the "catching-up" phase could be a separate entity in the architecture diagram and/or the code-base.

Even if what you propose might a neater diagram (I'm not convinced it does), there are practical realities that make such changes implausible.

Reply:
IIUC, "gossiping normally" is a push-style process where nodes push events to other nodes.

Again, IIUC, "catching-up phase" is a pull-style process where nodes request data from other nodes and then receive the data as responses to those requests.

How is push vs. pull model not a "binary change"?

Reply:
>IIUC, "gossiping normally" is a push-style process where nodes push events to other nodes. Again, IIUC, "catching-up phase" is a pull-style process where nodes request data from other nodes and then receive the data as responses to those requests.

That isn't how sync gossip works. A pair of nodes exchange information about what events each node has, followed by an exchange of events that each node is missing. Neither`push` nor `pull` meaningfully describe how the protocol works.

Reply:
Thanks for the clarification. If this codepath is indeed a part of "normal operations" indistinguishable from the regular gossip by any means, then I'd suggest to rephrase that part in the document to state that the non-expired events are crucial for normal gossip operations for any (slow-ish) node to be able to catch up with peers at any time, and not only after the reconnect.

Reply:
The updated wording looks great and super clear. Thank you!